smart-contracts/contracts/lib/SafeMath16.sol

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+// uint16 version of OpenZeppelin's SafeMath
+library SafeMath16 {
```

Comment:
Instead of type-specific implementations of `SafeMath` (`SafeMath16`, `SafeMath32`, ...) not originating from OpenZeppelin, we should use `SafeMath`(256bit) in combination with OpenZeppelin's `SafeCast`. Maintaining multiple math libraries that are not supported by `OpenZeppelin` would create for us an unnecessary maintenance burden and security weak point, especially when (eventually) updating the Solidity version. See https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1625#issuecomment-545978860

Reply:
I stumbled into this comment, but it seemed to add a lot of the burden to peform casting after every operation, but since this is kind of related to the next comment, I'll address it there.

Reply:
I stumbled upon this link when trying to figure out the best solution, but it seemed to add a lot of burden to perform this casting after every operation. Since this is related to the general `u256` comment, I'll address it further below.